### ISSUE TITLE
--wrap _Unwind_Resume doesn't work without -u _Unwind_Resume

### ISSUE THREAD

**Issue by smeenai** [llvm/llvm-project#50675]
|  |  |
| --- | --- |
| Bugzilla Link | [51333](https://llvm.org/bz51333) |
| Version | unspecified |
| OS | All |
| CC | @MaskRay,@lanza,@smithp35 |

## Extended Description 
```
unwind.s:

.globl _Unwind_Resume
.globl __wrap__Unwind_Resume
_Unwind_Resume:
__wrap__Unwind_Resume:
        retq

$ llvm-mc -triple x86_64-linux-gnu -filetype=obj -o unwind.o unwind.s
$ ld.lld -shared -o libunwind.so unwind.o

resume.ll:

target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"
define dso_local void @_Z1fv() optnone noinline personality i8* bitcast (void ()* @throw to i8*) {
  invoke void @throw()
          to label %unreachable unwind label %lpad
lpad:
  %1 = landingpad { i8*, i32 }
          cleanup
  resume { i8*, i32 } %1
unreachable:
  unreachable
}
declare void @throw()

$ llvm-as -o resume.bc resume.ll
$ ld.lld -shared -o libresume.so resume.bc libunwind.so --gc-sections --wrap _Unwind_Resume
```

Before https://github.com/llvm/llvm-project/commit/f96ff3c0f8ebd941b3f6b345164c3d858b781484, `__wrap__Unwind_Resume` is present in the dynamic symbol table for libunwind.so (as shown by running `nm -D libunwind.so`), as expected. After that commit, it's no longer present in the dynamic symbol table, and we have an invalid PLT entry instead.

`_Unwind_Resume` isn't present in the bitcode file's symbol table, since the reference to it is only synthesized during LTO codegen, so I imagine the symbol is considered unreferenced and the wrapper is incorrectly dead-stripped.

Linking against the DSO defining both `_Unwind_Resume` and `__wrap__Unwind_Resume` was necessary for the wrapping to work correctly even before that change, so there's possibly deeper underlying issues. We always happen to link against such a DSO though, so this was working for us prior to f96ff3c0f8eb.

Adding `-u _Unwind_Resume` works around the issue but is inelegant.

**Comment by smeenai** [comment]
We ran into this again in a different context (and it caused crashes for our users, which was unfortunate). https://reviews.llvm.org/D123985 makes us at least error out instead of producing a binary that'll crash at runtime. I also have an idea for a potential fix, which I'll put up separately.

**Comment by smeenai** [comment]
Should be fixed by 1af25a986069f2ae8c724133fa8649bb795a7925
